### PR TITLE
check available disk space before performing rook upgrade

### PR DESCRIPTION
**Pull request by laverya**
add ability for user to continue despite potentially lacking disk space

#### What this PR does / why we need it:

This PR estimates the amount of disk space required to upgrade Rook to the specified version and presents a warning if insufficient space is detected.

Both docker and containerd are supported, a limit of 85% is imposed on filesystems containing docker/containerd, and different logic is used if /var/lib/kurl is on the same filesystem as the image store.

An example message (when upgrading from Rook 1.0.4 to 1.10.8) is:
```bash
Would you like to continue? (y/N) y
Not enough disk space to upgrade Rook.
You need at least 78650 MB of free space on the filesystem containing /var/lib/kurl and /var/lib/containerd - and to have 15% free space after that to avoid image pruning.
Currently, only 18763 MB of free space is available before reaching 85% capacity.
If you have already loaded images or started this Rook upgrade, it is possible that less space will be required. Would you like to continue anyways?
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Warn the user if there is likely to be insufficient space to upgrade Rook many versions.
```

### CODE REVIEW

**Review by camilamacedo86 — APPROVED** [review]
It shows fine to me. 
The hard thing is that we cannot test it out prior get merged.

So, @laverya please run a testgrid afterwords with the scenarios to ensure that all is fine.

**Comment by emosbaugh** [comment]
What about secondary nodes? Would it be possible to add a similar check to

rook_upgrade_tasks_rook_upgrade

or

rook_upgrade_tasks_load_images

**Comment by laverya** [comment]
> ./scripts/common/rook-upgrade-test.sh: line 95: rook_upgrade_required_disk_space: command not found
> ASSERT:1.0 to 1.5 expected:<6100> but was:<>


🤦 I changed the function name

**Comment by laverya** [comment]
> What about secondary nodes? Would it be possible to add a similar check to
>
> rook_upgrade_tasks_rook_upgrade
>
>or
>
>rook_upgrade_tasks_load_images

yeah, that makes sense to do

File: scripts/common/rook-upgrade.sh [hunk 1]
```diff
@@ -95,6 +100,96 @@ function rook_upgrade_prompt() {
...
+        local available_mb=
+        available_mb="$((free_mb - total_mb * 3 / 20))" # free space, excluding 15% of the total
+
+        if [ "$available_mb" -lt "$total_required_size" ]; then
+            logWarn "Not enough disk space to upgrade Rook."
+            logWarn "You need at least $total_required_size MB of free space on the filesystem containing /var/lib/kurl and $container_directory - and to have 15%% free space after that to avoid image pruning."
+            logWarn "Currently, only $available_mb MB of free space is available before reaching 85%% capacity."
+            logWarn "If you have already loaded images or started this Rook upgrade, it is possible that less space will be required. Would you like to continue anyways?"
+            if ! confirmN; then
+                return 1
```
Comment (camilamacedo86) [inline scripts/common/rook-upgrade.sh:143]: Why we are not using bail here? WDYT about use bail?

File: scripts/common/rook-upgrade.sh [hunk 2]
```diff
@@ -95,6 +100,96 @@ function rook_upgrade_prompt() {
...
+            fi
+        fi
+    else
+        local kurl_dir_size=
+        kurl_dir_size=$((archive_size * 2))
+
+        local kurl_free_kb=
+        local kurl_free_mb=
+        kurl_free_kb="$(df -P /var/lib/kurl | awk 'END{print $4}')"
```
Comment (camilamacedo86) [inline scripts/common/rook-upgrade.sh:152]: I think we need to use the DIR here right? Otherwise, we are unable to test it from the source code.

Comment (emosbaugh) [reply scripts/common/rook-upgrade.sh:152]: Good call. Can we use the vars for this? `$BIN_KURL`

https://github.com/replicatedhq/kURL/blob/592f3ac1bcca78d91c3b7611fbc3b99752d4e40b/scripts/common/utilbinaries.sh#L8

Comment (laverya) [reply scripts/common/rook-upgrade.sh:152]: yeah, not hard to change

(just `$DIR` though)

File: scripts/common/rook-upgrade.sh (hunk 1 above)
Comment (laverya) [reply scripts/common/rook-upgrade.sh:143]: because for every other bit in this flow we're returning and not bailing

Comment (laverya) [reply scripts/common/rook-upgrade.sh:143]:
```
    if ! rook_upgrade_prompt "$current_version" "$desired_version" ; then
        bail "Not upgrading Rook"
    fi

    if ! rook_upgrade_storage_check "$current_version" "$desired_version" ; then
        bail "Not upgrading Rook"
    fi
```

Comment (camilamacedo86) [reply scripts/common/rook-upgrade.sh:143]: But in the `confirmN` we do not bail. Here:

```
if ! confirmN; then
                return 1
```

Should we not as we do in the other places?

Comment (laverya) [reply scripts/common/rook-upgrade.sh:143]: I can change it to just be `confirmN`, which is what we do in the `rook_upgrade_prompt` section

It's the same thing though, just spelled out

Comment (laverya) [reply scripts/common/rook-upgrade.sh:143]: `rook_upgrade_storage_check` returns 1, and so there is a bail in the snippet I pasted

Comment (camilamacedo86) [reply scripts/common/rook-upgrade.sh:143]: I see I think that is fine 👍 
Many ways to do the same :-)

**Review by camilamacedo86 — DISMISSED** [review]

**Review by camilamacedo86 — APPROVED** [review]

**Pull request merged**